### PR TITLE
Apache Thrift language binding for Dart

### PR DESCRIPTION
**Pull request by markerickson-wf**
IN PROGRESS

Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Currently basic functionality is implemented and the tutorial is working.  Integration tests will be written before a PR to Apache is made.

### CODE REVIEW

File: lib/dart/lib/src/browser/t_web_socket.dart [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        _onErrorController = new StreamController.broadcast(),
+        _onMessageController = new StreamController.broadcast() {
+    if (url == null || !url.hasAuthority || !url.hasPort) {
+      throw new ArgumentError("Invalid url");
+    }
+  }
+
+  WebSocket _socket;
+
+  bool get isOpen => _socket != null && _socket.readyState == WebSocket.OPEN;
```
Comment (evanweible-wf) [inline lib/dart/lib/src/browser/t_web_socket.dart:56]: Since you're requiring dart 1.12.0, you can use null-aware operators here:

``` dart
bool get isOpen => _socket?.readyState == WebSocket.OPEN;
```

File: lib/dart/lib/src/browser/t_web_socket.dart [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    _socket = new WebSocket(url.toString());
+    _socket.onError.listen(_onError);
+    _socket.onOpen.listen(_onOpen);
+    _socket.onClose.listen(_onClose);
+    _socket.onMessage.listen(_onMessage);
+  }
+
+  Future close() async {
+    if (_socket != null) {
+      _socket.close();
```
Comment (evanweible-wf) [inline lib/dart/lib/src/browser/t_web_socket.dart:78]: This method is async but you're not awaiting anything

File: lib/dart/lib/src/browser/t_web_socket.dart [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+    try {
+      data = CryptoUtils.base64StringToBytes(event.data);
+    } on FormatException catch (_) {
+      _onErrorController
+          .add(new UnsupportedError("Expected a Base 64 encoded string."));
+    }
+
+    if (!_completers.isEmpty) {
+      _completers.removeAt(0).complete(data);
```
Comment (evanweible-wf) [inline lib/dart/lib/src/browser/t_web_socket.dart]: Just to make sure I understand this correctly, is this assuming that every request sent to the server will be accompanied by a response? Is it actually guaranteed that they will come back in the same order? Or do they need to be mapped to an ID?

File: lib/dart/lib/src/browser/t_web_socket.dart [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+  void _onOpen(Event event) {
+    _onStateController.add(TSocketState.OPEN);
+    _sendRequests();
+  }
+
+  void _onClose(CloseEvent event) {
+    _socket = null;
+    _completers.clear();
+    _requests.clear();
```
Comment (evanweible-wf) [inline lib/dart/lib/src/browser/t_web_socket.dart:109]: Is it desirable to clear out pending requests when the socket closes? Would we ever need to know which requests didn't complete?

File: lib/dart/lib/src/protocol/t_json_protocol.dart [hunk 5]
```diff
@@ -0,0 +1,733 @@
...
+      byte += 10;
+      return byte - _Constants.HEX_A.codeUnitAt(0);
+    } else {
+      throw new TProtocolError(
+          TProtocolErrorType.INVALID_DATA, "Expected hex character");
+    }
+  }
+
+  int _hexChar(int byte) {
+    return byte.toRadixString(16).codeUnitAt(0);
```
Comment (evanweible-wf) [inline lib/dart/lib/src/protocol/t_json_protocol.dart]: `=>`

File: lib/dart/lib/src/transport/t_socket_transport.dart [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+  final bool isListening;
+
+  /// A transport using the provided [socket].  If [isListening] is true, then
+  /// messages received from [socket] will be written to the transport and made
+  /// available for reading.
+  TSocketTransport(this.socket, {isListening: false})
+      : this.isListening = isListening {
+    if (socket == null) {
+      throw new ArgumentError.notNull("socket");
```
Comment (evanweible-wf) [inline lib/dart/lib/src/transport/t_socket_transport.dart]: single quotes

File: tutorial/dart/client/pubspec.yaml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+name: tutorial_client
+version: 0.1.0
+description: A Dart client implementation of the Apache Thrift tutorial
+author: Mark Erickson <mark.erickson@workiva.com>
+homepage: https://github.com/apache/thrift
+
+environment:
+  sdk: '>=1.0.0'
```
Comment (evanweible-wf) [inline tutorial/dart/client/pubspec.yaml]: I read above that this requires dart >= 1.12.0, should that be reflected here?

Comment (evanweible-wf) [inline tutorial/dart/client/pubspec.yaml]: also, double quotes

File: tutorial/dart/client/pubspec.yaml [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+dependencies:
+  browser: '>=0.10.0'
```
Comment (evanweible-wf) [inline tutorial/dart/client/pubspec.yaml]: double quotes

File: lib/dart/lib/src/browser/t_web_socket.dart (hunk 3 above)
Comment (markerickson-wf) [inline lib/dart/lib/src/browser/t_web_socket.dart]: Yeah good catch it is assuming order, and for a client it assumes there will be a response.  This is actually what the JavaScript implementation does, but I had the same concerns when porting it over.

https://github.com/apache/thrift/blob/master/lib/js/src/thrift.js#L577

I think using the seqid would be preferable, but it seems that would require knowledge of the protocol being used to decode that, which the transport doesn't have.  We should think about this some more.

File: lib/dart/lib/src/browser/t_web_socket.dart (hunk 1 above)
Comment (markerickson-wf) [inline lib/dart/lib/src/browser/t_web_socket.dart:56]: I was holding off on those until tooling catches up and supports them.  But they will be nice!

File: lib/dart/lib/src/browser/t_web_socket.dart (hunk 2 above)
Comment (markerickson-wf) [inline lib/dart/lib/src/browser/t_web_socket.dart:78]: I'm using `async` to return the Future implicitly, to fulfill the interface.

File: lib/dart/lib/src/browser/t_web_socket.dart (hunk 4 above)
Comment (markerickson-wf) [inline lib/dart/lib/src/browser/t_web_socket.dart:109]: Good catch.  We should probably complete any remaining completers (both those in the active queue and those in requests) with an error before clearing.

File: lib/dart/lib/src/browser/t_web_socket.dart [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+  Future open() async {
+    if (!isClosed) {
+      throw new TTransportError(
+          TTransportErrorType.ALREADY_OPEN, 'Socket already connected');
+    }
+
+    _socket = new WebSocket(url.toString());
+    _socket.onError.listen(_onError);
+    _socket.onOpen.listen(_onOpen);
```
Comment (stevenosborne-wf) [inline lib/dart/lib/src/browser/t_web_socket.dart:69]: Presumably, we wouldn't want `open` to complete until `onOpen` completes. Could you return `_socket.onOpen.first`?

File: lib/dart/lib/src/browser/t_web_socket.dart (hunk 2 above)
Comment (stevenosborne-wf) [inline lib/dart/lib/src/browser/t_web_socket.dart:78]: Maybe return `_socket.onClose.first`.

**Comment by stevenosborne-wf** [comment]
First pass done. Looks great Mark!

File: lib/dart/lib/src/transport/t_buffered_transport.dart [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+/// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+/// KIND, either express or implied. See the License for the
+/// specific language governing permissions and limitations
+/// under the License.
+
+part of thrift;
+
+/// Buffered implementation of [TTransport].
+class TBufferedTransport extends TTransport {
+  final List<int> _writeBuffer = [];
```
Comment (tylertreat-wf) [inline lib/dart/lib/src/transport/t_buffered_transport.dart:22]: Is there a reason you use `List<int>` here instead of `Uint8List`?

**Comment by tylertreat-wf** [comment]
Got through most of the c++ code and it lgtm (though it's been a while since I've done any c++). Also went through the dart code and it looks good based on my primitive knowledge of dart.

File: lib/dart/lib/src/transport/t_buffered_transport.dart (hunk 10 above)
Comment (markerickson-wf) [inline lib/dart/lib/src/transport/t_buffered_transport.dart:22]: Uint8List is fixed length.  I need an internal buffer that is "growable".

File: lib/dart/lib/src/browser/t_web_socket.dart (hunk 9 above)
Comment (markerickson-wf) [inline lib/dart/lib/src/browser/t_web_socket.dart:69]: Good idea.  I addressed this in https://github.com/markerickson-wf/thrift/commit/cd83635f4ae688d2ae65d474790a213294576d15.

File: lib/dart/lib/src/browser/t_web_socket.dart (hunk 2 above)
Comment (markerickson-wf) [inline lib/dart/lib/src/browser/t_web_socket.dart:78]: Good idea.  I addressed this in https://github.com/markerickson-wf/thrift/commit/cd83635f4ae688d2ae65d474790a213294576d15.

**Comment by stevenosborne-wf** [comment]
+1

File: lib/dart/lib/src/browser/t_web_socket.dart [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    _socket.onClose.listen(_onClose);
+    _socket.onMessage.listen(_onMessage);
+
+    return _socket.onOpen.first;
```
Comment (evanweible-wf) [inline lib/dart/lib/src/browser/t_web_socket.dart:73]: Are these socket streams broadcast or single subscription? If they're not broadcast, this will throw since you're listening to onOpen above and here.

File: lib/dart/lib/src/t_base.dart [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+/// Unless required by applicable law or agreed to in writing,
+/// software distributed under the License is distributed on an
+/// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+/// KIND, either express or implied. See the License for the
+/// specific language governing permissions and limitations
+/// under the License.
+
+part of thrift;
+
+abstract class TBase {
```
Comment (johnryan-wf) [inline lib/dart/lib/src/t_base.dart:20]: is this being used?

File: lib/dart/lib/src/transport/t_transport.dart [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+      got += ret;
+    }
+    return got;
+  }
+
+  /// Writes the [bytes] to the output.
+  /// Throws [TTransportError] if there was an error writing data
+  void writeAll(Uint8List buffer) {
+    write(buffer, 0, buffer.length);
+  }
```
Comment (johnryan-wf) [inline lib/dart/lib/src/transport/t_transport.dart:65]: (nitpick) I would expect this method to be defined after `write()` (read, readAll, write, writeAll)

File: lib/dart/lib/src/browser/t_web_socket.dart (hunk 11 above)
Comment (markerickson-wf) [inline lib/dart/lib/src/browser/t_web_socket.dart:73]: `onOpen` and `onClose` for an html `WebSocket` return a `Stream` from `EventStreamProvider.forTarget()`, which is documented to return a broadcast stream. 

https://api.dartlang.org/1.12.1/dart-html/EventStreamProvider/forTarget.html

**Comment by johnryan-wf** [comment]
Looks great!  This is much appreciated!

File: lib/dart/lib/src/browser/t_web_socket.dart [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+
+  final StreamController<Object> _onErrorController;
+  Stream<Object> get onError => _onErrorController.stream;
+
+  final StreamController<Uint8List> _onMessageController;
+  Stream<Uint8List> get onMessage => _onMessageController.stream;
+
+  final List<Uint8List> _requests = [];
+
+  TWebSocket(this.url)
```
Comment (evanweible-wf) [inline lib/dart/lib/src/browser/t_web_socket.dart:45]: Seems a bit odd to me that the browser version expects a URL and the console version expects an instance of `dart:io.WebSocket`.

This also creates inconsistent behavior around `open()/close()` - as it is written now, the browser version could be reopened once closed since it instantiates a new underlying websocket, but the console version just assumes that the underlying websocket is open and adds that state to the state stream.

What is the goal for this socket class - should it be reopenable? Should it abstract the underlying websocket class away completely?

Comment (evanweible-wf) [inline lib/dart/lib/src/browser/t_web_socket.dart:45]: I see that the use case for creating a TWebSocket from an instance of WebSocket is in doing so from an upgrading request where you already have the websocket. It still seems that the default constructor for `TWebSocket` is different dependent on whether you import the browser or console version. This seems more like the console version should have an additional constructor since it can support being a websocket server and connecting to a websocket as a client.

Comment (markerickson-wf) [inline lib/dart/lib/src/browser/t_web_socket.dart:45]: That's a good observation.  I mainly implemented the console version to support running a server for the tutorial.  

I didn't finish a full TServer implementation (there is more to be done there), and I was a little uncertain as to how much time to invest on the server / console code, since that support is being left TODO.  I also don't have a good use case for a console app as a client, but I agree it could and probably should be supported.

Basically the question is whether to leave this as TODO or clean this up now.

File: lib/dart/lib/src/transport/t_transport.dart (hunk 13 above)
Comment (markerickson-wf) [inline lib/dart/lib/src/transport/t_transport.dart:65]: I agree, thanks.

File: lib/dart/lib/src/t_base.dart (hunk 12 above)
Comment (markerickson-wf) [inline lib/dart/lib/src/t_base.dart:20]: It's used by generated code.

File: lib/dart/lib/src/browser/t_web_socket.dart (hunk 14 above)
Comment (evanweible-wf) [inline lib/dart/lib/src/browser/t_web_socket.dart:45]: Ah I see, didn't realize there was a TServer interface. I'm fine with either

Comment (markerickson-wf) [inline lib/dart/lib/src/browser/t_web_socket.dart:45]: We'll leave this as TODO.

**Comment by markerickson-wf** [comment]
PR has been made to Apache - https://github.com/apache/thrift/pull/608

**Comment by johnryan-wf** [comment]
:+1:
